HousingResidentInformationAPI.Tests/V1/UseCase/ValidatePostcodeTests.cs

```diff
@@ -22,6 +22,11 @@ public void ValidPostcodesReturnTrue(string postcode)
         }
 
         [TestCase("1")]
+        [TestCase("111111")]
+        [TestCase("111 111")]
+        [TestCase("EEEEEE")]
```

Comment:
Another thing I noticed about your [Regex](https://github.com/LBHackney-IT/housing-resident-information-api/blob/d9ea99c598620947e99a493a387470612414349a/HousingResidentInformationAPI/V1/UseCase/ValidatePostcode.cs#L11-L12):
`^([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9]?[A-Za-z])))) [0-9][A-Za-z]{2})$`

is that your 6th `([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})` and 9th `([A-Za-z][A-Ha-hJ-Yj-y][0-9]?[A-Za-z])` capture groups aren't tested when it comes to the character limitations that they impose on the Outcode. So, for instance, if you tested "EI9 9LL", then it should fail, because your 6th capture group won't allow character "I" in the Outcode. It won't get detected, though because of the bug in your Regex.

Reply:
The regex pattern is copied from documented use, from a government source. Probably taking into account non obvious facts about postcodes.
Do you have a better alternative?

Reply:
> The regex pattern is copied from documented use, from a government source. Probably taking into account non obvious facts about postcodes.

Yes it does - for instance this part: `[A-Ha-hJ-Yj-y]` disallows characters "I" and "Z" by design. And it's a code you that you have untested. If you choose to include this part, you should also test this part.

According to our standards and even this APIs Tests part of the [README](https://github.com/LBHackney-IT/housing-resident-information-api/blob/master/README.md#testing). This should have been done in the TDD way, I think at the very least you could try not to violate: "Test coverage should never go down" to make for it, by including tests for these parts of the Regex that you chose to have - otherwise you have untested code, which in turn reduces the test coverage of an API.

I don't agree with the mindset of: `"I copied it from the internet, it probably works, hence there's no need too test it fully"`. I'm not willing to give in when it comes to tests, hence I'm unresolving this conversation. 😐

Reply:
In order to add these tests you'll need too fix #52 bug, though.

Just because Regex in some official Pdf document, doesn't mean it will necessarily work - you need to test these things for yourself. I tested it in practice in dotnet fiddle - it's broken - it's a fact. In fact it's broken so much so that this `[A-Ha-hJ-Yj-y]` part of the Regex ceases to do it's job:

![image](https://user-images.githubusercontent.com/43747286/101982242-5c705e00-3c6a-11eb-9076-de9427252517.png)

This doesn't look like a valid postcode for me!

Reply:
I get a sense that you don't care about having your code tested and bug free as much as I do.
We can agree to disagree on this - it's all opinion, and you can request approval from another reviewer, who could agree with you on this matter.

Reply:
What is currently not valid about that postcode in the image? It likely could become a future postcode? I'm not a postcode expert. 
Please can you provide an improved regex pattern?
You've not suggested an alternative or better regex pattern like I asked previously.  Just criticism on a regex pattern widely recognised, documented and used. 
Where did I say this: "I copied it from the internet, it probably works, hence there's no need too test it fully".  That's a massive assumption to make from my initial reply.
I'm not sure this is a progressive discussion at all.

Reply:
Yes it is an assumption - it's extrapolated from this:

> Although not perfect/exhaustive I would prefer to stick with this regex pattern, I presume it comes from a source that  has tested it well before documenting it for public use.
 Source being:  https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/488478/Bulk_Data_Transfer_-_additional_validation_valid_from_12_November_2015.pdf.

> The regex pattern is copied from documented use, from a government source. Probably taking into account non obvious facts about postcodes.

From #52 
> ABCDEFEA7 8HJ or abcdefea7 8hj is a false positive now, but maybe in the future it isn't it?

and from that you're arguing against adding more test cases, especially when I point out the postcodes that would fail your tests:
- `ABCDEFEA7 8HJ` validator should return **false**,  but returns **true**.
- `EI9 9LL` validator should return **false**, but returns **true**

Reply:
> I'm not sure this is a progressive discussion at all.

Am afraid it seems so... . I would also want to say that me writing: `"I copied it from the internet, it probably works, hence there's no need too test it fully"` in no way was intended to affect you in a negative way.

The reason I wrote that was to show a simplified impression I got out of the conversation, and to say that I disagree with that.

Reply:
I'm actually doing a PR to improve the test coverage and that is your assumption.
Please refer to https://stackoverflow.com/questions/164979/regex-for-matching-uk-postcodes .
The regex isn't perfect or exhaustive, but it is widely used. The false positives maybe are flaws? 
Please provide something you think is an improvement.

Reply:
I did that yesterday on #52. Don't know whether you look at `Potential fixes:` section, but it contains a fixed expression.